src/Access/DiskAccessStorage.cpp

```diff
@@ -660,58 +546,32 @@ bool DiskAccessStorage::updateImpl(const UUID & id, const UpdateFunc & update_fu
 
 bool DiskAccessStorage::updateNoLock(const UUID & id, const UpdateFunc & update_func, bool throw_if_not_exists, bool write_on_disk)
 {
-    auto it = entries_by_id.find(id);
-    if (it == entries_by_id.end())
+    AccessEntityPtr old_entity = memory_storage.read(id, throw_if_not_exists);
+    if (!old_entity)
     {
         if (throw_if_not_exists)
             throwNotFound(id);
         else
             return false;
     }
 
-    Entry & entry = it->second;
     if (readonly)
-        throwReadonlyCannotUpdate(entry.type, entry.name);
-
-    if (!entry.entity)
-        entry.entity = readAccessEntityFromDisk(id);
-    auto old_entity = entry.entity;
-    auto new_entity = update_func(old_entity, id);
-
-    if (!new_entity->isTypeOf(old_entity->getType()))
-        throwBadCast(id, new_entity->getType(), new_entity->getName(), old_entity->getType());
-
-    if (*new_entity == *old_entity)
-        return true;
+        throwReadonlyCannotUpdate(old_entity->getType(), old_entity->getName());
 
-    const String & new_name = new_entity->getName();
-    const String & old_name = old_entity->getName();
-    const AccessEntityType type = entry.type;
-    auto & entries_by_name = entries_by_name_and_type[static_cast<size_t>(type)];
-
-    bool name_changed = (new_name != old_name);
-    if (name_changed)
-    {
-        if (entries_by_name.contains(new_name))
-            throwNameCollisionCannotRename(type, old_name, new_name);
-        if (write_on_disk)
-            scheduleWriteLists(type);
+    if (isNotLoadedFromDisk(old_entity)) {
+        old_entity = readAccessEntityFromDisk(id);
+        memory_storage.insert(id, old_entity, /* replace_if_exists= */ true, /* throw_if_exists= */ false, /* conflicting_id= */ nullptr);
     }
 
+    if (!memory_storage.update(id, update_func, throw_if_not_exists))
+        return false;
+
+    AccessEntityPtr new_entity = memory_storage.read(id, throw_if_not_exists);
+    if (old_entity->getName() != new_entity->getName() && write_on_disk)
+        scheduleWriteLists(new_entity->getType());
     if (write_on_disk)
         writeAccessEntityToDisk(id, *new_entity);
```

Comment:
These conditions look strange, but that's how it was in the original implementation. Maybe it should be under one if.

Reply:
Looks good to me: if the access entity name changed - we write a new list to the disk. And we also always write the access entity itself. Can be simplified a little bit, though: 
```
if (write_on_disk)
{
    if (old_entity->getName() != new_entity->getName())
        scheduleWriteLists(new_entity->getType());
     writeAccessEntityToDisk(id, *new_entity);
}
```